test/ssl/client_ssl_authentication_test.go

```diff
@@ -0,0 +1,245 @@
+	if err != nil {
+		return "", err
+	}
+	remoteController.StartMember(cluster.ID)
+	return cluster.ID, nil
+}
+
+func createClientConfigWithSSLConfig(clientCertPath string, clientKeyPath string, caPath string) (*config.Config, error) {
+	config := hazelcast.NewConfig()
+	sslConfig := config.NetworkConfig().SSLConfig()
+	sslConfig.SetEnabled(true)
+	err := sslConfig.SetCaPath(caPath)
+	if err != nil {
+		return nil, err
+	}
+	err = sslConfig.AddClientCertAndKeyPath(clientCertPath, clientKeyPath)
+	if err != nil {
+		return nil, err
+	}
+	sslConfig.ServerName = serverName
+	return config, nil
+}
+
+func TestSSLMutualAuthenticationConnect(t *testing.T) {
+	if !test.IsEnterprise() {
+		t.Skipf("SSL feature requires enterprise version")
+	}
+	clusterID, err := createMemberWithXML(maRequiredXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig(client1Cert,
+		client1Key, server1CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	client, err := hazelcast.NewClientWithConfig(config)
+	defer client.Shutdown()
+
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestSSLMutualAuthentication_ClientDoesntKnowServerFail(t *testing.T) {
+	if !test.IsEnterprise() {
+		t.Skipf("SSL feature requires enterprise version")
+	}
+	clusterID, err := createMemberWithXML(maRequiredXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig(client1Cert,
+		client1Key, server2CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = hazelcast.NewClientWithConfig(config)
+	if _, ok := err.(*core.HazelcastIllegalStateError); !ok {
+		t.Error(err)
+	}
+}
+
+func TestSSLMutualAuthentication_ServerDoesntKnowClientFail(t *testing.T) {
+	if !test.IsEnterprise() {
+		t.Skipf("SSL feature requires enterprise version")
+	}
+	clusterID, err := createMemberWithXML(maRequiredXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig(client2Cert,
+		client2Key, server2CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = hazelcast.NewClientWithConfig(config)
+	if _, ok := err.(*core.HazelcastIllegalStateError); !ok {
+		t.Error(err)
+	}
+}
+
+func TestSSLMutualAuthentication_NeitherServerNotClientKnowsTheOtherFail(t *testing.T) {
+	if !test.IsEnterprise() {
+		t.Skipf("SSL feature requires enterprise version")
+	}
+	clusterID, err := createMemberWithXML(maRequiredXML)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer remoteController.ShutdownCluster(clusterID)
+	config, err := createClientConfigWithSSLConfig("client2-cert.pem",
+		client2Key, server2CA)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = hazelcast.NewClientWithConfig(config)
+	if _, ok := err.(*core.HazelcastIllegalStateError); !ok {
+		t.Error(err)
+	}
+}
+
+func TestSSLOptionalMutualAuthenticationConnect(t *testing.T) {
+	if !test.IsEnterprise() {
```

Comment:
Why do you need this check in the tests? I assume that you always need to use the ee jar for server during your tests, what is the exception to this?